cuda_bindings/cuda/bindings/_path_finder/load_dl_windows.py

```diff
@@ -0,0 +1,130 @@
+    os.add_dll_directory(dirpath)
+    # Update PATH as a fallback for dependent DLL resolution
+    curr_path = os.environ.get("PATH")
+    os.environ["PATH"] = dirpath if curr_path is None else os.pathsep.join((curr_path, dirpath))
+
+
+def abs_path_for_dynamic_library(handle: int) -> str:
+    """Get the absolute path of a loaded dynamic library on Windows.
+
+    Args:
+        handle: The library handle
+
+    Returns:
+        The absolute path to the DLL file
+
+    Raises:
+        OSError: If GetModuleFileNameW fails
+    """
+    buf = ctypes.create_unicode_buffer(260)
+    n_chars = ctypes.windll.kernel32.GetModuleFileNameW(ctypes.wintypes.HMODULE(handle), buf, len(buf))
+    if n_chars == 0:
+        raise OSError("GetModuleFileNameW failed")
+    return buf.value
+
+
+def check_if_already_loaded_from_elsewhere(libname: str) -> Optional[LoadedDL]:
+    """Check if the library is already loaded in the process.
+
+    Args:
+        libname: The name of the library to check
+
+    Returns:
+        A LoadedDL object if the library is already loaded, None otherwise
+
+    Example:
+        >>> loaded = check_if_already_loaded_from_elsewhere("cudart")
+        >>> if loaded is not None:
+        ...     print(f"Library already loaded from {loaded.abs_path}")
+    """
+    from .supported_libs import SUPPORTED_WINDOWS_DLLS
+
+    for dll_name in SUPPORTED_WINDOWS_DLLS.get(libname, ()):
+        try:
+            handle = win32api.GetModuleHandle(dll_name)
+        except pywintypes.error:
+            continue
+        else:
+            return LoadedDL(handle, abs_path_for_dynamic_library(handle), True)
+    return None
+
+
+def load_with_system_search(name: str, _unused: str) -> Optional[LoadedDL]:
+    """Try to load a DLL using system search paths.
+
+    Args:
+        name: The name of the library to load
+        _unused: Unused parameter (kept for interface consistency)
+
+    Returns:
+        A LoadedDL object if successful, None if the library cannot be loaded
+    """
+    from .supported_libs import SUPPORTED_WINDOWS_DLLS
+
+    dll_names = SUPPORTED_WINDOWS_DLLS.get(name)
+    if dll_names is None:
+        return None
+
+    for dll_name in dll_names:
+        handle = ctypes.windll.kernel32.LoadLibraryW(ctypes.c_wchar_p(dll_name))
+        if handle:
+            return LoadedDL(handle, abs_path_for_dynamic_library(handle), False)
+
+    return None
+
+
+def load_with_abs_path(libname: str, found_path: str) -> LoadedDL:
+    """Load a dynamic library from the given path.
+
+    Args:
+        libname: The name of the library to load
+        found_path: The absolute path to the DLL file
+
+    Returns:
+        A LoadedDL object representing the loaded library
+
+    Raises:
+        RuntimeError: If the DLL cannot be loaded
+    """
+    from .supported_libs import LIBNAMES_REQUIRING_OS_ADD_DLL_DIRECTORY
+
+    if libname in LIBNAMES_REQUIRING_OS_ADD_DLL_DIRECTORY:
+        add_dll_directory(found_path)
```

Comment:
Q: this is needed only when loading with abs path? Other cases are fine without it?

Reply:
Good question...

We have the five cases in README.md:

0. already loaded → that's obviously fine
1. sys.path / site-packages → produces `found_path`, which is covered here
2. conda → dito
3. system → dito
4. OS Default → that's the last resort, we have exhausted all other options, if this doesn't work (unexpectedly) we have no way to succeed with the load

Does that seem right?

Reply:
Sounds about right to me! Thanks.